odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,259 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import datetime
+import itertools
+import json
+import logging
+import re
+import traceback
+import threading
+import odoo
+import odoo.exceptions
+from odoo import models, fields, api
+from odoo.http import serialize_exception
+
+_logger = logging.getLogger(__name__)
+BASE_VERSION = odoo.modules.load_information_from_description_file('base')['version']
+
+
+class IrAsync(models.Model):
+    _name = 'ir.async'
+    _description = 'Asynchronous Jobs'
+
+    name = fields.Char()
+    state = fields.Selection([
+        ('created', 'Created'),     # The job has been enqueued for later processing
+        ('succeeded', 'Succeeded'), # The job finished with a result, the result has not yet been retrieved
+        ('failed', 'Failed'),       # The job failed due to an exception
+        ('done', 'Done'),           # The job finished without result or the result has been retrieved
+    ])
+    res_model = fields.Char()
+    res_ids = fields.Text()
+    method = fields.Char()
+    args = fields.Text()
+    kwargs = fields.Text()
+    user_id = fields.Many2one('res.users', string='Scheduler User', ondelete='cascade')
+    context = fields.Text()
+    super_user = fields.Boolean()
+    traceback = fields.Text()
+    payload = fields.Text(default=False)
+
+    def call(self, target, args=None, kwargs=None):
+        """
+        Thread/Process-like API to create asynchronous jobs.
+
+        The method will be called later as ``target(*args, **kwargs)`` in a
+        new transaction using a copy of the current environment (user, context,
+        recordset).  Arguments must be serializable in JSON.
+
+        The job's return value is stored in JSON in the ``payload`` field, and
+        web-notification are possible via bus/models/ir_async:IrAsync.call_notify
+        """
+        recs = getattr(target, '__self__', None)
+        if not isinstance(recs, models.BaseModel):
+            raise TypeError("You can only create an async task on a recordset")
+        model = recs.__class__
+        context = recs.env.context.copy()
+        context.pop('async_traceback', '')
+        context.pop('async_job_id', '')
+
+        job = self.sudo().create({
+            'state': 'created',
+            'res_model': model._name,
+            'res_ids': json.dumps(recs._ids),
+            'method': target.__name__,
+            'args': json.dumps(args or []),
+            'kwargs': json.dumps(kwargs or {}),
+            'user_id': int(recs.env.uid),
+            'context': json.dumps(recs.env.context),
+            'super_user': recs.env.su,
+            'traceback': self._merge_traceback(traceback.format_stack()[:-1]),
+        })
+
+        self._cr.after("commit", self._notifydb)
+
+        return job
+
+    def call_notify(self, target, args=None, kwargs=None, description=False):
```

Comment:
If you go for the `partial`-like API, the description must come first
```suggestion
    def call_notify(self, description, method, *args, **kwargs):
```